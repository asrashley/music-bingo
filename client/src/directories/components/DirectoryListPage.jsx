import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { DetailsPanel } from './DetailsPanel';

import {
  clearSeachResults,
  fetchDirectoriesIfNeeded,
  fetchDirectoryDetailIfNeeded,
  toggleDirectoryExpand,
  searchForSongs
} from '../directoriesSlice';

import { fetchUserIfNeeded } from '../../user/userSlice';

import {
  getDirectoryList,
  getDirectoryMap,
  getSortOptions,
  getLastUpdated,
  getLocation,
  getSearchResults,
  getSearchText,
  getIsSearching
} from '../directoriesSelectors';

import { getUser } from '../../user/userSelectors';

import '../styles/directories.scss';

const SongRow = ({ onSelect, depth, song, selected }) => {
  let className = "song";
  if (selected.song?.pk === song.pk) {
    className += ' is-selected';
  }
  return (
    <tr className={className}>
      <td className="select"></td>
      <td className="title" style={{ paddingLeft: `${depth}em` }}>
        <button className="btn btn-link song-link" onClick={(ev) => { ev.preventDefault(); onSelect({ song }); }} >
          {song.title}
        </button>
      </td>
      <td className="artist">
        {song.artist}
      </td>
    </tr>
  );
};

SongRow.propTypes = {
  song: PropTypes.object.isRequired
};

const SongListing = ({ depth, songs, onSelect, selected }) => {
  return (
    <React.Fragment>
      {songs.map((song, idx) => (<SongRow depth={depth} song={song} onSelect={onSelect}
        key={idx} selected={selected} />))}
    </React.Fragment>
  );
};

SongListing.propTypes = {
  options: PropTypes.object.isRequired,
  onSelect: PropTypes.func.isRequired,
  songs: PropTypes.array.isRequired,
  selected: PropTypes.object.isRequired
};

const TitleCell = ({ className, directory, onSelect, onVisibilityToggle }) => {
  if (directory.directories.length > 0 || directory.songs.length > 0) {
    return (<button className={`${className} btn btn-link directory-link`} onClick={(ev) => onVisibilityToggle(directory)}>
      {directory.title}</button>);
  }
  return (<button
    className={`${className} btn btn-link directory-link leaf`}
    onClick={(ev) => { ev.preventDefault(); onSelect({ directory }); }}
  >{directory.title}</button>);
};

TitleCell.propTypes = {
  options: PropTypes.object.isRequired,
  directory: PropTypes.object.isRequired,
  onVisibilityToggle: PropTypes.func.isRequired
};

const DirectoryChildren = ({ depth, directory, options, onSelect, onVisibilityToggle, selected }) => {
  return (
    <React.Fragment>
      {
        directory.directories.length > 0 && <DirectoryListing
          options={options}
          directories={directory.directories}
          depth={depth}
          onSelect={onSelect}
          onVisibilityToggle={onVisibilityToggle}
          selected={selected} />
      }
      {
        (directory.songs.length > 0 && directory.valid === true) && <SongListing options={options}
          depth={depth + 1} songs={directory.songs} onSelect={onSelect} selected={selected} />
      }
    </React.Fragment>
  );
};

const DirectoryRow = ({ options, depth, directory, onSelect, onVisibilityToggle, selected }) => {
  const expandable = directory.directories.length > 0 || directory.songs.length > 0;
  const titleClass = selected.directory?.pk === directory.pk ? 'title is-selected' : 'title';
  const prefix = directory.expanded ? '-' : '+';
  let className = 'directory';
  if (directory.expanded) {
    className += ' expanded';
  } else if (expandable) {
    className += ' expandable';
  }
  return (
    <React.Fragment>
      <tr className={className}>
        <td className="select">
          <button className="btn btn-outline-dark btn-sm" onClick={(ev) => onVisibilityToggle(directory)}>{prefix}</button>
        </td>
        <td className="directory" colSpan="2" style={{ paddingLeft: `${depth}em` }}>
          <TitleCell
            className={titleClass}
            directory={directory}
            onSelect={onSelect}
            options={options}
            onVisibilityToggle={onVisibilityToggle}
          />
        </td>
      </tr>
      {
        expandable && directory.expanded && <DirectoryChildren
          depth={depth + 1}
          directory={directory}
          options={options}
          onSelect={onSelect}
          onVisibilityToggle={onVisibilityToggle}
          selected={selected}
        />
      }
    </React.Fragment>
  );
};

const DirectoryListing = ({ depth, directories, options, onSelect, onVisibilityToggle, selected }) => {
  return <React.Fragment>
    {
      directories.map((dir, idx) => (
        <DirectoryRow
          depth={depth}
          directory={dir}
          options={options}
          key={idx}
          onSelect={onSelect}
          onVisibilityToggle={onVisibilityToggle}
          selected={selected}
        />))
    }
  </React.Fragment>;
};

function SearchResults({ onSelect, options, queryResults, selected }) {
  return (
    <table className="search-results table table-bordered table-sm table-hover">
      <caption>Search results</caption>
      <thead className="thead-light">
        <tr>
          <th className="select"></th>
          <th className="title">Title</th>
          <th className="artist">Artist</th>
        </tr>
      </thead>
      <tbody>
        <SongListing
          options={options}
          onSelect={onSelect}
          songs={queryResults}
          selected={selected} />
      </tbody>
    </table>
  );
}

class DirectoryListPage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      ActiveDialog: null,
      dialogData: null,
      query: "",
      selected: {
        directory: null,
        song: null
      }
    };
    this.timeout = null;
  }

  componentDidMount() {
    const { directoryMap, dispatch, location, user } = this.props;
    dispatch(clearSeachResults());
    dispatch(fetchUserIfNeeded());
    if (user.pk > 0) {
      dispatch(fetchDirectoriesIfNeeded());
      if (location !== undefined) {
        if (directoryMap[location]?.expanded === false) {
          dispatch(fetchDirectoryDetailIfNeeded(location));
          dispatch(toggleDirectoryExpand({ dirPk: location }));
        }
      }
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { directoryMap, dispatch, location, user } = this.props;
    if (prevProps.user.pk !== user.pk && user.pk > 0) {
      dispatch(fetchDirectoriesIfNeeded());
      if (location !== undefined) {
        if (directoryMap[location]?.expanded === false) {
          dispatch(fetchDirectoryDetailIfNeeded(location));
          dispatch(toggleDirectoryExpand({ dirPk: location }));
        }
      }
    }
    if (user.pk > 0 && location !== undefined &&
      prevProps.directories.length !== this.props.directories.length) {
      if (directoryMap[location]?.expanded === false) {
        dispatch(fetchDirectoryDetailIfNeeded(location));
        dispatch(toggleDirectoryExpand({ dirPk: location }));
      }
    }
  }

  componentWillUnmount() {
    this.clearTimeout();
  }

  clearTimeout = () => {
    if (this.timeout !== null) {
      window.clearTimeout(this.timeout);
      this.timeout = null;
    }
  };

  onVisibilityToggle = (directory) => {
    const { pk, expanded } = directory;
    const { dispatch } = this.props;
    if (!expanded) {
      dispatch(fetchDirectoryDetailIfNeeded(pk));
    }
    dispatch(toggleDirectoryExpand({ dirPk: pk }));
    this.setState({ selected: { directory, song: null } });
  };

  onSelect = ({ directory, song }) => {
    if (directory === undefined) {
      directory = null;
    }
    if (song === undefined) {
      song = null;
    }
    this.setState({ selected: { directory, song } });
  };

  onChangeQuery = (event) => {
    const { location, searchText } = this.props;
    const minLength = location ? 1 : 3;
    const timeout = location ? 100 : 250;
    event.preventDefault();
    this.clearTimeout();
    this.setState({ query: event.target.value });
    if (event.target.value.length >= minLength ||
      (searchText !== '' && searchText !== event.target.value)) {
      this.timeout = setTimeout(this.makeQuery, timeout);
    }
  };

  makeQuery = () => {
    const { dispatch, isSearching, location } = this.props;
    const { query } = this.state;
    const minLength = location ? 1 : 3;
    if (isSearching === true) {
      this.clearTimeout();
      this.timeout = setTimeout(this.makeQuery, 100);
      return;
    }
    this.setState({
      selected: {
        directory: null,
        song: null
      }
    });
    if (query.length < minLength) {
      dispatch(clearSeachResults());
    } else {
      dispatch(searchForSongs(query, location));
    }
  };

  render() {
    const { directories, directoryMap, location, options, user, queryResults } = this.props;
    const { ActiveDialog, dialogData, selected, query } = this.state;
    let column1;

    if (location === undefined && queryResults.length > 0) {
      column1 = <SearchResults
        options={options}
        onSelect={this.onSelect}
        queryResults={queryResults}
        selected={selected} />;
    } else {
      column1 = <table className="directory-listing table table-sm table-bordered table-hover">
        <thead className="thead-light">
          <tr>
            <th className="select"></th>
            <th className="title">Title</th>
            <th className="artist">Artist</th>
          </tr>
        </thead>
        <tbody>
          <DirectoryListing
            className="directory-listing"
            depth={0}
            directories={directories}
            options={options}
            onSelect={this.onSelect}
            onVisibilityToggle={this.onVisibilityToggle}
            selected={selected}
          />
        </tbody>
      </table>;
    }
    return (
      <div id="directories-page" className={user.loggedIn ? '' : 'modal-open'}  >
        <input
          name="query"
          placeholder="search..."
          type="text"
          className="song-search"
          value={query}
          onChange={this.onChangeQuery}
        />
        <div className="column1">{column1}</div>
        <DetailsPanel className="directory-detail" selected={selected} directoryMap={directoryMap} />
        {ActiveDialog && <ActiveDialog backdrop {...dialogData} />}
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    directories: getDirectoryList(state, ownProps),
    directoryMap: getDirectoryMap(state, ownProps),
    isSearching: getIsSearching(state, ownProps),
    options: getSortOptions(state, ownProps),
    lastUpdated: getLastUpdated(state, ownProps),
    location: getLocation(state, ownProps),
    queryResults: getSearchResults(state, ownProps),
    searchText: getSearchText(state, ownProps),
    user: getUser(state, ownProps),
  };
};

DirectoryListPage = connect(mapStateToProps)(DirectoryListPage);

export { DirectoryListPage };
