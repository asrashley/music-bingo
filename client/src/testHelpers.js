import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import log from 'loglevel';
import { screen, fireEvent } from '@testing-library/react';
import waitForExpect from 'wait-for-expect';

import { DisplayDialog } from './components/DisplayDialog';
import { createStore } from './store/createStore';
import { history } from './store/history';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (<Provider store={store}>
      <ConnectedRouter history={history}>
        <DisplayDialog>
          {children}
        </DisplayDialog>
      </ConnectedRouter>
    </Provider>);
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

const protectedRoutes = {
  '/api/directory': true,
  '/api/game': true,
  '/api/settings': false,
  '/api/user': true,
};

export function jsonResponse(payload, status = 200) {
  const body = JSON.stringify(payload);
  return {
    body,
    status,
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': body.length
    }
  };
};

export function installFetchMocks(fetchMock, {
  loggedIn = false,
  currentAccessToken = 1,
  refreshToken = "refresh.token"
} = {}) {
  const responseModifiers = {};
  let serverStatus = null;
  const apiRequest = async (url, opts) => {
    log.trace(`apiRequest ${url}`);
    if (serverStatus !== null) {
      return serverStatus;
    }
    if (protectedRoutes[url]) {
      const { headers } = opts;
      log.trace(`Authorization = headers?.Authorization`);
      const bearer = `Bearer ${getAccessToken()}`;
      if (headers?.Authorization !== bearer) {
        log.debug(`Invalid access token "${headers?.Authorization}" !== "${bearer}"`);
        return 401;
      }
    }
    const filename = url.replace(/^\/api/, './fixtures');
    log.trace(`load ${filename}`);
    const data = await import(`${filename}.json`);
    log.trace(`${filename} = ${Object.keys(data['default']).join(',')}`);
    if (url === '/api/settings' && !loggedIn) {
      data['default'] = {
        privacy: data['default'].privacy
      };
    }
    const modFn = responseModifiers[url];
    if (modFn !== undefined) {
      return jsonResponse(modFn(url, data['default']));
    }
    return jsonResponse(data['default']);
  };
  const getAccessToken = () => `access.token.${currentAccessToken}`;
  const refreshAccessToken = (url, opts) => {
    currentAccessToken++;
    log.trace(`refreshAccessToken ${currentAccessToken}`);
    return {
      'accessToken': getAccessToken()
    };
  };
  const checkUser = async (url, opts) => {
    log.debug(`checkUser loggedIn=${loggedIn}`);
    if (serverStatus !== null) {
      log.debug(`checkUser serverStatus=${serverStatus}`);
      return serverStatus;
    }
    if (!loggedIn) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    return jsonResponse(data['default']);
  };
  const userDatabase = [{
    username: 'user',
    email: 'a.user@example.tld',
    password: 'mysecret'
  }];
  const loginUser = async (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`loginUser status=${serverStatus}`);
      return serverStatus;
    }
    const { username, password } = JSON.parse(opts.body);
    const userIsValid = userDatabase.some(item => (
      item.username === username && item.password === password));
    log.debug(`loginUser: username="${username}" password="${password}" valid=${userIsValid}`);
    if (!userIsValid) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    loggedIn = true;
    return jsonResponse(data['default']);
  };
  const logoutUser = (url, opts) => {
    if (serverStatus !== null) {
      log.trace(`logoutUser status=${serverStatus}`);
      return serverStatus;
    }
    log.trace('logoutUser');
    loggedIn = false;
    return jsonResponse('Logged out');
  };
  const checkIfUserExists = (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`checkIfUserExists status=${serverStatus}`);
      return serverStatus;
    }
    const { username, email } = JSON.parse(opts.body);
    const response = {
      "username": false,
      "email": false
    };
    userDatabase.forEach((item) => {
      if (item.username === username) {
        response.username = true;
      }
      if (item.email === email) {
        response.email = true;
      }
    });
    log.debug(`checkIfUserExists username=${username} email=${email} response=${JSON.stringify(response)}`);
    return jsonResponse(response);
  };

  fetchMock.config.fallbackToNetwork = false;
  fetchMock.config.warnOnFallback = true;
  log.debug(`installFetchMocks ${loggedIn}`);
  fetchMock
    .get('/api/directory', apiRequest)
    .get(/api\/directory\/\d+/, apiRequest)
    .get('/api/games', apiRequest)
    .get(/api\/game\/\d+/, apiRequest)
    .post('/api/refresh', refreshAccessToken)
    .get('/api/settings', apiRequest)
    .get('/api/user', checkUser)
    .post('/api/user', loginUser)
    .delete('/api/user', logoutUser)
    .post('/api/user/check', checkIfUserExists)
    .get('/api/users', apiRequest)
    .get('/api/user/guest', apiRequest);

  return {
    getAccessToken,
    getRefreshToken: () => refreshToken,
    isLoggedIn: () => loggedIn,
    addUser: (user) => userDatabase.push(user),
    setResponseModifier: (url, fn) => {
      responseModifiers[url] = fn;
    },
    setServerStatus: (code) => {
      serverStatus = code;
    },
    jsonResponse
  };
}

export async function setFormFields(fields) {
  fields.forEach(({ label, value, exact }) => {
    if (exact === undefined) {
      exact = true;
    }
    fireEvent.input(screen.getByLabelText(label, { exact }), {
      target: {
        value
      }
    });
  });
  await waitForExpect(() => {
    const last = fields[fields.length - 1];
    const { value } = document.querySelector(`input[name="${last.label}"`);
    expect(value).toBe(fields[fields.length - 1].value);
  });
}

