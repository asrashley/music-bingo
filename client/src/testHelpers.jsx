import { render, queries, buildQueries, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import { ReduxRouter } from '@lagunovsky/redux-react-router';
import log from 'loglevel';
import { Readable } from "stream";
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';

import { DisplayDialog } from './components/DisplayDialog';
import { createStore } from './store/createStore';
import fetchMockHandlerTest from 'fetch-mock/esm/client';
import jestify from 'fetch-mock-jest/jestify';

import userData from './fixtures/user.json';

export const fetchMock = jestify(fetchMockHandlerTest);

/* custom query that looks for a "data-last-update" attribute in the container and compares that to
  the value provided to the query. It can be used to check if a component has been updated or
  to wait for it to re-render
  */
const queryAllByLastUpdate = (container, lastUpdate, options = {}) => {
  const { comparison } = {
    comparison: 'equals',
    ...options
  };
  log.trace(`queryAllByLastUpdate lastUpdate=${lastUpdate} comparison=${comparison} container="${container.nodeName}.${container.className}" ${container.id}`);
  return Array.from(container.querySelectorAll('[data-last-update]'))
    .filter((elt) => {
      const update = parseInt(elt.dataset.lastUpdate, 10);
      log.trace(`${elt.nodeName}.${elt.className} lastUpdate = ${update}`);
      if (isNaN(update)) {
        return false;
      }
      switch (comparison) {
        case 'equals':
          return update === lastUpdate;
        case 'greaterThan':
          return update > lastUpdate;
        case 'greaterThanOrEquals':
          return update >= lastUpdate;
        case 'lessThan':
          return update < lastUpdate;
        case 'lessThanOrEquals':
          return update <= lastUpdate;
        default:
          return true;
      }
    });
};
export const [queryByLastUpdate, getAllLastUpdate, getByLastUpdate, findAllLastUpdate, findByLastUpdate] = buildQueries(
  queryAllByLastUpdate,
  (container, selector) => `Found multiple elements from ${container} with last update selector: ${selector}`,
  (container, selector, opts = {}) => {
    const { comparison = 'equals' } = opts;
    return (`Unable to find an element from ${container.nodeName}.${container.className} with last update ${comparison} ${selector}`);
  }
);
const lastUpdatedQueries = { queryByLastUpdate, getAllLastUpdate, getByLastUpdate, findAllLastUpdate, findByLastUpdate };

/* custom query that uses a CSS selector to find elements */
const queryAllBySelector = (container, selector) => {
  log.trace(`queryAllBySelector ${selector}`);
  return Array.from(container.querySelectorAll(selector));
};
const [queryBySelector, getAllBySelector, getBySelector, findAllBySelector, findBySelector] = buildQueries(
  queryAllBySelector,
  (container, selector) => `Found multiple elements from ${container} with selector: ${selector}`,
  (container, selector) => `Unable to find an element from ${container} with selector: ${selector}`,
);
const bySelectorQueries = { queryBySelector, getAllBySelector, getBySelector, findAllBySelector, findBySelector };

const routerSelector = (state) => state.router

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    history = createMemoryHistory(),
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (<Provider store={store}>
      <ReduxRouter history={history} routerSelector={routerSelector}>
        <DisplayDialog>
          {children}
        </DisplayDialog>
      </ReduxRouter>
    </Provider>);
  }
  Wrapper.propTypes = {
    children: PropTypes.node,
  };
  const events = userEvent.setup({
    delay: 10,
  });
  return {
    events,
    store,
    ...render(ui, {
      wrapper: Wrapper,
      queries: {
        ...queries,
        ...lastUpdatedQueries,
        ...bySelectorQueries
      },
      ...renderOptions
    })
  };
}

const protectedRoutes = {
  '/api/directory': true,
  '/api/game': true,
  '/api/settings': false,
  '/api/user': true,
};

export function jsonResponse(payload, status = 200) {
  const body = JSON.stringify(payload);
  return {
    body,
    status,
    headers: {
      'Cache-Control': 'max-age = 0, no_cache, no_store, must_revalidate',
      'Content-Type': 'application/json',
      'Content-Length': body.length
    }
  };
}

export class MockResponse extends Response {
  constructor(...args) {
    if (args[0] instanceof ReadableStream) {
      args[0] = Readable.from(args[0]);
    }
    super(...args);
  }
}

export function installFetchMocks(fetchMock, {
  loggedIn = false,
  currentAccessToken = 1,
  refreshToken = "refresh.token"
} = {}) {
  const responseModifiers = {};
  let serverStatus = null;
  const apiRequest = async (url, opts) => {
    log.trace(`apiRequest ${url}`);
    if (serverStatus !== null) {
      return serverStatus;
    }
    if (protectedRoutes[url]) {
      const { headers } = opts;
      log.trace(`Authorization = headers?.Authorization`);
      const bearer = `Bearer ${getAccessToken()}`;
      if (headers?.Authorization !== bearer) {
        log.debug(`Invalid access token "${headers?.Authorization}" !== "${bearer}"`);
        return 401;
      }
    }
    const filename = url.replace(/^\/api/, './fixtures');
    log.trace(`load ${filename}`);
    const data = await import(`${filename}.json`);
    log.trace(`${filename} = ${Object.keys(data['default']).join(',')}`);
    if (url === '/api/settings' && !loggedIn) {
      data['default'] = {
        privacy: data['default'].privacy
      };
    }
    const modFn = responseModifiers[url];
    let responsePayload = data['default'];
    if (modFn !== undefined) {
      log.debug(`modify response for ${url}`);
      responsePayload = modFn(url, responsePayload);
      log.debug(`modify response for ${url} done`);
    }
    return jsonResponse(responsePayload);
  };
  const getAccessToken = () => `access.token.${currentAccessToken}`;
  const refreshAccessToken = () => {
    currentAccessToken++;
    log.trace(`refreshAccessToken ${currentAccessToken}`);
    return {
      'accessToken': getAccessToken()
    };
  };
  const checkUser = async () => {
    log.debug(`checkUser loggedIn=${loggedIn}`);
    if (serverStatus !== null) {
      log.debug(`checkUser serverStatus=${serverStatus}`);
      return serverStatus;
    }
    if (!loggedIn) {
      return 401;
    }
    return jsonResponse({
      ...userData,
      refreshToken,
      accessToken: getAccessToken(),
    });
  };
  const userDatabase = [{
    username: 'user',
    email: 'a.user@example.tld',
    password: 'mysecret'
  }];
  const loginUser = async (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`loginUser status=${serverStatus}`);
      return serverStatus;
    }
    const { username, password } = JSON.parse(opts.body);
    let index = -1;
    let dbEntry;
    userDatabase.forEach((item, i) => {
      if (item.username === username && item.password === password) {
        dbEntry = item;
        index = i;
      }
    });
    log.debug(`loginUser: username="${username}" password="${password}" index=${index}`);
    if (!dbEntry) {
      return 401;
    }
    const user = {
      ...userData,
      pk: index + 2,
      ...dbEntry,
      refreshToken,
      accessToken: getAccessToken()
    };
    delete user.password;
    loggedIn = true;
    return jsonResponse(user);
  };
  const logoutUser = () => {
    if (serverStatus !== null) {
      log.trace(`logoutUser status=${serverStatus}`);
      return serverStatus;
    }
    log.trace('logoutUser');
    loggedIn = false;
    return jsonResponse('Logged out');
  };
  const checkIfUserExists = (url, opts) => {
    if (serverStatus !== null) {
      log.trace(`checkIfUserExists status=${serverStatus}`);
      return serverStatus;
    }
    const { username, email } = JSON.parse(opts.body);
    const response = {
      "username": false,
      "email": false
    };
    userDatabase.forEach((item) => {
      if (item.username === username) {
        response.username = true;
      }
      if (item.email === email) {
        response.email = true;
      }
    });
    log.debug(`checkIfUserExists username=${username} email=${email} response=${JSON.stringify(response)}`);
    return jsonResponse(response);
  };

  Object.assign(fetchMock.config, {
    fallbackToNetwork: false,
    warnOnFallback: true,
    Response: MockResponse,
  });
  log.trace(`installFetchMocks ${loggedIn}`);
  fetchMock
    .get('/api/directory', apiRequest)
    .get(/api\/directory\/\d+/, apiRequest)
    .get('/api/games', apiRequest)
    .get(/api\/game\/\d+/, apiRequest)
    .post('/api/refresh', refreshAccessToken)
    .get('/api/settings', apiRequest)
    .get('/api/user', checkUser)
    .post('/api/user', loginUser)
    .delete('/api/user', logoutUser)
    .post('/api/user/check', checkIfUserExists)
    .get('/api/users', apiRequest)
    .get('/api/user/guest', apiRequest);

  return {
    getAccessToken,
    getRefreshToken: () => refreshToken,
    isLoggedIn: () => loggedIn,
    addUser: (user) => userDatabase.push(user),
    logout: () => loggedIn = false,
    setResponseModifier: (url, fn) => {
      responseModifiers[url] = fn;
    },
    setServerStatus: (code) => {
      serverStatus = code;
    },
    jsonResponse
  };
}

export async function setFormFields(fields, events) {
  await Promise.all(fields.map(async ({ label, value, exact }) => {
    if (exact === undefined) {
      exact = true;
    }
    const elt = await screen.findByLabelText(label, { exact });
    if (events) {
      //await events.click(elt);
      await waitFor(() => events.clear(elt));
      /*
            await waitFor(() => events.clear(elt));
        await waitFor(() => events.type(elt, value));
      }*/
    }
    await waitFor(() => fireEvent.input(elt, {
      target: {
        value
      }
    }));
  }));
  const { label, exact, value } = fields[fields.length - 1];
  const elt = await screen.findByLabelText(label, { exact });
  await waitFor(() => {
    expect(elt.value).toBe(value);
  });
}

