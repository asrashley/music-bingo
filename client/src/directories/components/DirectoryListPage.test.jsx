import React from 'react';
import log from 'loglevel';
import fetchMock from "fetch-mock-jest";
import { screen } from '@testing-library/react';

import { renderWithProviders, installFetchMocks } from '../../testHelpers';
import { DirectoryListPage } from './DirectoryListPage';
//import { initialState } from '../../store/initialState';

//import { DirectoryInitialState } from '../directoriesSlice';

describe('DirectoryListPage component', () => {
	beforeEach(() => {
		installFetchMocks(fetchMock, { loggedIn: true });
	});
	afterEach(() => {
		fetchMock.mockReset();
		log.resetLevel();
	});
	it('renders a directory', async () => {
		const location = {
				params: {
					dirPk: 1
				}
		};
		//log.setLevel('debug');
		const { asFragment } = renderWithProviders(<DirectoryListPage match={location} /> /*, { preloadedState }*/);
		await screen.findByText("100 Nows");
		expect(asFragment()).toMatchSnapshot();
	});
	afterEach(() => log.resetLevel());
});