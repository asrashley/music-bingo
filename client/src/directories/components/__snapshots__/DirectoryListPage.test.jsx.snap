// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectoryListPage component renders a directory 1`] = `
<DocumentFragment>
  <div
    class=""
    id="directories-page"
  >
    <input
      class="song-search"
      name="query"
      placeholder="search..."
      type="text"
      value=""
    />
    <div
      class="column1"
    >
      <table
        class="directory-listing table table-sm table-bordered table-hover"
      >
        <thead
          class="thead-light"
        >
          <tr>
            <th
              class="select"
            />
            <th
              class="title"
            >
              Title
            </th>
            <th
              class="artist"
            >
              Artist
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="directory expanded"
          >
            <td
              class="select"
            >
              <button
                class="btn btn-outline-dark btn-sm"
              >
                -
              </button>
            </td>
            <td
              class="directory"
              colspan="2"
              style="padding-left: 0em;"
            >
              <button
                class="title btn btn-link directory-link"
              >
                Clips
              </button>
            </td>
          </tr>
          <tr
            class="directory expandable"
          >
            <td
              class="select"
            >
              <button
                class="btn btn-outline-dark btn-sm"
              >
                +
              </button>
            </td>
            <td
              class="directory"
              colspan="2"
              style="padding-left: 1em;"
            >
              <button
                class="title btn btn-link directory-link"
              >
                100 Nows
              </button>
            </td>
          </tr>
          <tr
            class="directory expandable"
          >
            <td
              class="select"
            >
              <button
                class="btn btn-outline-dark btn-sm"
              >
                +
              </button>
            </td>
            <td
              class="directory"
              colspan="2"
              style="padding-left: 1em;"
            >
              <button
                class="title btn btn-link directory-link"
              >
                2000s
              </button>
            </td>
          </tr>
          <tr
            class="directory expandable"
          >
            <td
              class="select"
            >
              <button
                class="btn btn-outline-dark btn-sm"
              >
                +
              </button>
            </td>
            <td
              class="directory"
              colspan="2"
              style="padding-left: 1em;"
            >
              <button
                class="title btn btn-link directory-link"
              >
                2010s
              </button>
            </td>
          </tr>
          <tr
            class="directory expandable"
          >
            <td
              class="select"
            >
              <button
                class="btn btn-outline-dark btn-sm"
              >
                +
              </button>
            </td>
            <td
              class="directory"
              colspan="2"
              style="padding-left: 1em;"
            >
              <button
                class="title btn btn-link directory-link"
              >
                American
              </button>
            </td>
          </tr>
          <tr
            class="directory expandable"
          >
            <td
              class="select"
            >
              <button
                class="btn btn-outline-dark btn-sm"
              >
                +
              </button>
            </td>
            <td
              class="directory"
              colspan="2"
              style="padding-left: 1em;"
            >
              <button
                class="title btn btn-link directory-link"
              >
                Chilled Christmas
              </button>
            </td>
          </tr>
          <tr
            class="directory expandable"
          >
            <td
              class="select"
            >
              <button
                class="btn btn-outline-dark btn-sm"
              >
                +
              </button>
            </td>
            <td
              class="directory"
              colspan="2"
              style="padding-left: 1em;"
            >
              <button
                class="title btn btn-link directory-link"
              >
                Disco
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="directory-detail"
    />
  </div>
</DocumentFragment>
`;
